### PR TITLE
🐛 fix: implement p256verify precompile (RIP-7212)

### LINKED ISSUE #1906: bug: p256verify Precompile (RIP-7212) Returns Incorrect/Empty Value
## Description

Calling the p256verify precompile at address 0x...0100 results in an empty 0x return value. According to RIP-7212, it should return a 32-byte padded 1 for a valid signature or 0 for any failure, and consume a fixed gas amount.

This appears to just be an unimplemented feature in ethereumjs as there is a draft pr that never got merged before ethjs team got disbanded.

## Prompt

<prompt_document>
    <metadata>
        <title>Definitive Prompt: Implementing the RIP-7212 p256verify Precompile in TEVM</title>
        <version>1.0</version>
        <objective>
            To guide an expert TypeScript developer in implementing the `p256verify` precompile (as defined in RIP-7212) into the `tevm` library. The primary goals are to fix a bug causing an incorrect return value and to integrate the precompile as a default, overridable feature within the `createTevmNode` factory function.
        </objective>
    </metadata>

    <role_definition>
        You are an expert TypeScript developer specializing in Ethereum Virtual Machine (EVM) implementations and advanced cryptography. You write clean, maintainable, and robust code, adhering strictly to a Test-Driven Development (TDD) methodology. Your work is precise and accounts for all specified edge cases.
    </role_definition>

    <problem_statement>
        <summary>
            Currently, calling the `p256verify` precompile address (`0x00...0100`) within `tevm` yields an incorrect result (`0x` empty) instead of the expected success value (`0x00...01`). This indicates a missing or flawed implementation of the precompile.
        </summary>
        <technical_details>
            The precompile at address `0x0000000000000000000000000000000000000100` is expected to perform an ECDSA signature verification on the secp256r1 (P256) curve. A successful verification should return a 32-byte `Uint8Array` representing the value `1`. Any failure, including invalid signatures or malformed input, should return a 32-byte `Uint8Array` representing `0`. The gas cost for this operation is fixed at `3450`, regardless of the outcome.
        </technical_details>
        <bug_reproduction_script>
            <description>The following script demonstrates the bug and should serve as the foundation for the primary success-case test. It currently produces an empty `returnValue`.</description>
            <code>
            ```typescript
            import { base } from "@tevm/common";
            import { createEvm } from "@tevm/evm";
            import { createStateManager } from "@tevm/state";
            import { createAddress } from "tevm/address";
            import { createChain } from "tevm/blockchain";
            import { hexToBytes, bytesToHex } from "viem";

            const evm = await createEvm({
              common: base,
              stateManager: createStateManager({}),
              blockchain: await createChain({ common: base }),
              customPrecompiles: [],
            });

            const result = await evm.runCall({
              caller: createAddress("0x2a6c7bb649234ee2656550e163c8aaaed7318dcb"), // This value does not affect the outcome
              to: createAddress("0x0000000000000000000000000000000000000100"), // RIP-7212 Address
              data: hexToBytes(
                "c74ace4c2ccdb912b6876fa178a4a7adb6ea0916bfa73aa2c73fb4df5ce133a6" + // r
                "ae85d3657b170fb227cd404e3ae80e1974e885d6c0999094aad732979040be80" + // s
                "2c795862878f462f200a403b062c1b24e7de207f0c16f3e4d98d4c221c5e653b" + // x
                "2bd4817b59b8bdc0157af76bd95077d68a96c53a15c84fbd568c8759364aa1bf" + // y
                "e928602caf3f7716ee83abc596147665d9adfe7154a05440555571cefbe9652c"   // msgHash
              ),
            });

            console.log(bytesToHex(result.execResult.returnValue)); // Expected: 0x00...01
            console.log(result.execResult.executionGasUsed); // Expected: 3450n
            ```
            </code>
        </bug_reproduction_script>
    </problem_statement>

    <implementation_requirements>
        <success_criteria>
            - **Bug Resolution:** The `p256verify` precompile at `0x...0100` correctly returns `pad(toBytes(1), { size: 32 })` for valid inputs and `pad(toBytes(0), { size: 32 })` for all failure conditions.
            - **TDD Adherence:** The implementation is demonstrably test-driven, with a clear commit/PR history showing the red-green-refactor cycle.
            - **Default Integration:** `createTevmNode()` loads the `p256verify` precompile by default without requiring any user configuration.
            - **User Overridability:** A user can provide their own `customPrecompiles` array that successfully overrides the default `p256verify` precompile if an address conflicts.
            - **Gas Accuracy:** The precompile consistently consumes `3450n` gas per execution, regardless of success, failure, or malformed input.
            - **Comprehensive Test Coverage:** All new code has 100% test coverage for valid inputs, invalid signatures, incorrect input lengths, and malformed data. All existing project tests continue to pass.
        </success_criteria>

        <testing_philosophy>
            ### Testing Philosophy
            🚨 **CRITICAL: Follow a strict TDD approach - write tests first, implement second, refactor third.**

            **TDD Workflow:**
            - **Red:** Write failing tests for expected behavior.
            - **Green:** Implement the minimal amount of code required to make the tests pass.
            - **Refactor:** Optimize and improve the code structure while keeping all tests green.
            - **Repeat:** Apply this cycle for each new requirement or edge case.
        </testing_philosophy>

        <development_workflow>
            <step name="Step 1: Write Failing Tests (Red)">
                Create `packages/precompiles/test/p256verify.spec.ts`. Write a `vitest` test suite using `createTevmNode` that covers:
                1.  A **valid signature** case using the `bug_reproduction_script` data, asserting the correct return value (`1`) and gas (`3450n`).
                2.  An **invalid signature** case (e.g., with a flipped bit in `s`), asserting the correct return value (`0`) and gas (`3450n`).
                3.  An **incorrect input length** case (e.g., 159 bytes), asserting the correct return value (`0`) and gas (`3450n`).
                4.  A **malformed data** case (e.g., all-zero input), asserting the correct return value (`0`) and gas (`3450n`).
                All these tests must fail initially because the precompile is not yet implemented or integrated.
            </step>
            <step name="Step 2: Implement the Precompile Logic (Green)">
                Create `packages/precompiles/src/p256verify.precompile.ts`. Implement the precompile function logic as specified in the `<technical_specifications>` section below. The goal is to make all tests from Step 1 pass.
            </step>
            <step name="Step 3: Integrate and Write Integration Tests">
                Modify `packages/node/src/createTevmNode.ts` to include the precompile as a default. Then, in `packages/node/test/createTevmNode.spec.ts`, add a new test suite:
                1.  Verify the precompile is loaded and works correctly by default when calling `createTevmNode()` with no options.
                2.  Verify the user override mechanism works by passing a mock precompile at the same address and asserting it is used instead of the default.
            </step>
            <step name="Step 4: Refactor">
                Review all new code. Refactor for clarity, remove any code duplication, and add comments where necessary. Ensure the entire project's test suite passes.
            </step>
        </development_workflow>

        <technical_specifications>
            <input_validation>
                If the input `data` length is **not exactly 160 bytes**, the precompile MUST return `pad(toBytes(0), { size: 32 })` and consume the full `3450n` gas. This check must happen before any other processing.
            </input_validation>
            <data_parsing>
                The 160-byte input `data` buffer must be parsed as follows:
                - **r:** bytes 0-32
                - **s:** bytes 32-64
                - **x:** bytes 64-96
                - **y:** bytes 96-128
                - **msgHash:** bytes 128-160
            </data_parsing>
            <cryptographic_tooling>
                - **Dependency:** You **must** use the `@noble/curves` library.
                - **Import:** `import { p256 } from '@noble/curves/nist';`
                - **Key Construction:** The public key for verification must be constructed internally as a 65-byte uncompressed key: `Uint8Array([0x04, ...x, ...y])`.
                - **Signature Construction:** The signature for verification must be a 64-byte compact `Uint8Array` of `(r, s)`.
                - **Verification Call:** Use `p256.verify(signature, msgHash, publicKey)` within a `try...catch` block. Any exception caught during verification must result in a failure (`0`) return value.
            </cryptographic_tooling>
            <gas_accounting>
                The `executionGasUsed` MUST always be `3450n`, regardless of the outcome.
            </gas_accounting>
            <return_values>
                - **Success:** `pad(toBytes(1), { size: 32 })`
                - **Failure (any reason):** `pad(toBytes(0), { size: 32 })`
            </return_values>
            <integration_logic>
                - In `createTevmNode.ts`, you must merge the default `p256verify` precompile with any `options.customPrecompiles`. User-provided precompiles for the same address MUST override the default.
                - **Address:** The precompile address `0x00...0100` **must** be created using `new EthjsAddress(hexToBytes(...))`.
                - **Merge Logic:**
                    ```typescript
                    // Pseudo-code for merging logic
                    const defaultPrecompiles = [p256Precompile];
                    const userPrecompiles = options.customPrecompiles ?? [];
                    const userAddresses = new Set(userPrecompiles.map(p => p.address.toString()));
                    const filteredDefaults = defaultPrecompiles.filter(p => !userAddresses.has(p.address.toString()));
                    const finalPrecompiles = [...filteredDefaults, ...userPrecompiles];
                    ```
            </integration_logic>
        </technical_specifications>
    </implementation_requirements>

    <reference_materials>
        <reference_implementations>
            <summary>Reference implementations in Geth and REVM provide strong conventions to follow.</summary>
            <implementation source="go-ethereum">
                <description>Geth establishes patterns for how precompiles should handle errors and validate input.</description>
                - **Precompile Definitions:** `core/vm/contracts.go` - [View on GitHub](https://github.com/ethereum/go-ethereum/blob/master/core/vm/contracts.go)
                - **Key Convention:** Precompiles rarely throw errors. Instead, they return a padded zero value and still consume the full gas cost, even on malformed input. Input length is strictly validated.
            </implementation>
            <implementation source="revm">
                <description>REVM provides a modern, robust reference in Rust.</description>
                - **Location:** `crates/precompile/src/secp256r1.rs`
                - **Key Convention:** Always charges `3450` gas. Strictly validates input is 160 bytes. Returns a padded `1` on success and an empty/zero value on failure. Uses extensive test vectors.
            </implementation>
        </reference_implementations>

        <dependency_documentation library="@noble/curves">
            <description>
            The P256 curve, also known as **NIST secp256r1**, is a widely used elliptic curve specified by NIST for cryptographic operations. The `@noble/curves` library provides a fast, audited, and easy-to-use JavaScript/TypeScript implementation.
            </description>
            - **Installation:** `npm install @noble/curves`
            - **Importing:** `import { p256 } from '@noble/curves/nist.js';`
            - **API:**
                - `p256.verify(signature, msgHash, publicKey)`: Verifies an ECDSA signature.
                - `signature`: 64-byte compact `Uint8Array` of `(r, s)`.
                - `msgHash`: 32-byte `Uint8Array`.
                - `publicKey`: 65-byte uncompressed `Uint8Array`.
        </dependency_documentation>

        <project_context>
            <description>The following file content is provided as crucial context for the integration step.</description>
            <file_to_modify file_path="packages/node/src/createTevmNode.ts">
                <code>
                ```typescript
                import { createChain } from '@tevm/blockchain'
                import { createCommon, tevmDefault } from '@tevm/common'
                // ... (rest of the createTevmNode.ts file content as provided in the conversation)
                // ...
                const evmPromise = Promise.all([chainCommonPromise, stateManagerPromise, blockchainPromise]).then(
                    ([common, stateManager, blockchain]) => {
                        // THIS IS WHERE THE PRECOMPILE LOGIC WILL BE INTEGRATED
                        return createEvm({
                            common,
                            stateManager,
                            blockchain,
                            allowUnlimitedContractSize: options.allowUnlimitedContractSize ?? false,
                            customPrecompiles: options.customPrecompiles ?? [], // This line will be modified
                            profiler: options.profiler ?? false,
                            loggingLevel,
                        })
                    },
                )
                // ... (rest of the file)
                ```
                </code>
            </file_to_modify>
        </project_context>
    </reference_materials>

    <deliverables>
        <description>Upon completion, you must provide the following four code artifacts.</description>
        1.  **`p256verify.precompile.ts`** (New File): The file containing the standalone precompile logic.
        2.  **`p256verify.spec.ts`** (New File): The test file proving the precompile's logic, including all edge cases.
        3.  **`createTevmNode.ts`** (Modified File): The updated file showing the default integration logic.
        4.  **`createTevmNode.spec.ts`** (Modified File): The updated test file verifying the default loading and user override mechanisms.
    </deliverables>
</prompt_document>

### PR DESCRIPTION
**Pull request by roninjin10**
Fixes #1906

Implements the missing p256verify precompile as specified in RIP-7212 to resolve the bug where calling the precompile at address 0x...0100 returned an empty value instead of the expected 32-byte result.

## Changes

- Add @noble/curves dependency for secp256r1 signature verification
- Implement p256verify precompile with proper input validation and gas accounting
- Integrate precompile as default in createTevmNode with user override capability
- Add comprehensive test suite covering all edge cases
- Fixed gas cost to 3450n and correct return values per RIP-7212 spec

## Test Plan

- [x] Valid signature verification returns padded 1
- [x] Invalid signature verification returns padded 0
- [x] Incorrect input length returns padded 0
- [x] Malformed data returns padded 0
- [x] All cases consume exactly 3450n gas
- [x] Default integration in createTevmNode
- [x] User override capability works correctly

🤖 Generated with [Claude Code](https://claude.ai/code)

### CODE REVIEW

File: packages/node/src/createTevmNode.js [hunk 1]
```diff
@@ -230,12 +231,27 @@ export const createTevmNode = (options = {}) => {
...
+			const userPrecompiles = options.customPrecompiles ?? []
+			
+			// Create a set of user precompile addresses for efficient lookup
+			const userAddresses = new Set(userPrecompiles.map(p => p.address.toString()))
+			
+			// Filter out any default precompiles that are overridden by user precompiles
+			const filteredDefaults = defaultPrecompiles.filter(p => !userAddresses.has(p.address.toString()))
+			
+			// Merge filtered defaults with user precompiles
+			const finalPrecompiles = [...filteredDefaults, ...userPrecompiles]
```
Comment (roninjin10) [inline packages/node/src/createTevmNode.js:247]: @claude remove these comments please and inline stuff more don't create so many variables

File: packages/precompiles/package.json [hunk 2]
```diff
@@ -61,6 +61,7 @@
     "test:run": "vitest"
   },
   "dependencies": {
+    "@noble/curves": "^1.6.0",
```
Comment (roninjin10) [inline packages/precompiles/package.json:64]: @claude please use npm to verify this is the latest version of noble/curves. If it is just mention it is if not please update to latest

File: packages/precompiles/src/index.ts [hunk 3]
```diff
@@ -4,3 +4,4 @@ export type { TypedError } from './TypedError.js'
 export type { ConstructorArgument } from './ConstructorArgument.js'
 export { definePrecompile } from './definePrecompile.js'
 export { defineCall } from './defineCall.js'
+export { createP256VerifyPrecompile, P256_VERIFY_ADDRESS, P256_VERIFY_GAS_COST } from './p256verify.precompile.js'
```
Comment (roninjin10) [inline packages/precompiles/src/index.ts:7]: @claude Let's rename this to just p256VerifyPrecompile

**Review by roninjin10 — COMMENTED** [review]
@claude I have reviewed your pr. Please add a test to memory-client package as well
